lib/luacompletion.c

```diff
@@ -0,0 +1,88 @@
+/*
+* SPDX-FileCopyrightText: (c) 2023-2024 Ring Zero Desenvolvimento de Software LTDA
+* SPDX-License-Identifier: MIT OR GPL-2.0-only
+*/
+
+#define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/completion.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <lunatik.h>
+
+LUNATIK_OBJECTCHECKER(luacompletion_check, struct completion *);
+
+static int luacompletion_complete(lua_State *L)
+{
+	struct completion *completion = luacompletion_check(L, 1);
+
+	complete(completion);
+	return 0;
+}
+
+static int luacompletion_wait(lua_State *L)
+{
+	struct completion *completion = luacompletion_check(L, 1);
+	lua_Integer timeout = luaL_optinteger(L, 2, MAX_SCHEDULE_TIMEOUT);
+	unsigned long timeout_jiffies = msecs_to_jiffies((unsigned long)timeout);
+	unsigned long ret;
+
+	ret = wait_for_completion_interruptible_timeout(completion, timeout_jiffies);
+	if (ret == 0 || ret == -ERESTARTSYS) {
```

Comment:
I would handle other negative cases as "unknown".. just for "completion" (no pun intended 😂).. in this case, I would first check for `> 0` and pushbool/return 1, then I would switch-case with a default: "unknown".. just a suggestion though ;-)

Reply:
it could be, for the sake of completeness, in case the internal API changes. will change.

Reply:
I missed something important. `wait_for_completion_interruptible_timeout` returns an `unsigned long`; if I were to check for anything other than the specific known return values, I'd have to cast it to signed. That might be too much of an assumption, don't you think?

Reply:
sorry, I made a confusion here with the previous function we were using, the prototypes are actually: 
- `unsigned long wait_for_completion_timeout(struct completion *done, unsigned long timeout)`
- `long wait_for_completion_interruptible_timeout(struct completion *done, unsigned long timeout)`